flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,201 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.connector.base.source.hybrid;
+
+import org.apache.flink.api.connector.source.ReaderOutput;
+import org.apache.flink.api.connector.source.SourceEvent;
+import org.apache.flink.api.connector.source.SourceReader;
+import org.apache.flink.api.connector.source.SourceReaderContext;
+import org.apache.flink.api.connector.source.SourceSplit;
+import org.apache.flink.core.io.InputStatus;
+import org.apache.flink.util.Preconditions;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/** Hybrid source reader that delegates to the actual current source reader. */
+public class HybridSourceReader<T, SplitT extends SourceSplit>
+        implements SourceReader<T, HybridSourceSplit<SplitT>> {
+    private static final Logger LOG = LoggerFactory.getLogger(HybridSourceReader.class);
+    private static final int SOURCE_READER_FINISHED_EVENT_DELAY = 250;
```

Comment:
Could the value of `SOURCE_READER_FINISHED_EVENT_DELAY ` be configured by users? And why does this need the `SOURCE_READER_FINISHED_EVENT_DELAY`?

Reply:
Delay is required to not flood the coordinator with finished events when the current reader reaches end of input. Yes, it can be made configurable.

Reply:
I am not sure this induced jittering is really necessary. there should be some natural jitters on when readers reaching end of input.

Reply:
This is used within poll. I added the delay after I realized that too many events were generated. The reason for that is that the reader converts the underlying `END_OF_INPUT` to `MORE_AVAILABLE`. But more input will only be available after the reader was switched. I will check if this issue would go away when using `NOTHING_AVAILABLE` instead.

Reply:
The delay is no longer required with https://github.com/apache/flink/pull/15924/commits/3a01cd27d1fb36e1fa7faf5500bf97617ce22ed2